src/backend/distributed/progress/multi_progress.c

```diff
@@ -204,24 +215,38 @@ ProgressMonitorData *
 MonitorDataFromDSMHandle(dsm_handle dsmHandle, dsm_segment **attachedSegment)
 {
 	dsm_segment *dsmSegment = dsm_find_mapping(dsmHandle);
-	ProgressMonitorData *monitor = NULL;
 
 	if (dsmSegment == NULL)
 	{
 		dsmSegment = dsm_attach(dsmHandle);
 	}
 
-	if (dsmSegment != NULL)
+	if (dsmSegment == NULL)
 	{
-		monitor = (ProgressMonitorData *) dsm_segment_address(dsmSegment);
-		monitor->steps = (void *) (monitor + 1);
-		*attachedSegment = dsmSegment;
+		return NULL;
 	}
 
+	ProgressMonitorData *monitor = (ProgressMonitorData *) dsm_segment_address(
+		dsmSegment);
+
+	*attachedSegment = dsmSegment;
+
 	return monitor;
 }
 
 
+/*
+ * ProgressMonitorSteps returns a pointer to the array of steps that are stored
+ * in a progress monitor. This is simply the data right after the header, so
+ * this function is trivial.
+ */
+void *
+ProgressMonitorSteps(ProgressMonitorData *monitor)
+{
+	return monitor + 1;
```

Comment:
shouldn't this be something like `&(monitor->steps)`?

Reply:
No, because I removed the `steps` field in the struct completely, it now only contains the `stepsCount` and the `processId`. The header is `ProgressMonitorData`. Right after this header, there is memory allocated for `monitor->stepsCount * stepSize`. By doing the `+ 1`, we advance the pointer to right after the header, which is where the "array" of steps begins.

Reply:
> I removed the steps field in the struct completely

Ah, I missed that. Makes sense.